connmgr/error.go

```diff
@@ -4,60 +4,100 @@
 	// the configuration.
-	ErrDialNil = Err("ErrDialNil")
+	ErrDialNil = ErrorKind("ErrDialNil")
 
 	// ErrBothDialsFilled is used to indicate that Dial and DialAddr
 	// cannot both be specified in the configuration.
-	ErrBothDialsFilled = Err("ErrBothDialsFilled")
+	ErrBothDialsFilled = ErrorKind("ErrBothDialsFilled")
 
 	// ErrTorInvalidAddressResponse indicates an invalid address was
 	// returned by the Tor DNS resolver.
-	ErrTorInvalidAddressResponse = Err("ErrTorInvalidAddressResponse")
+	ErrTorInvalidAddressResponse = ErrorKind("ErrTorInvalidAddressResponse")
 
 	// ErrTorInvalidProxyResponse indicates the Tor proxy returned a
 	// response in an unexpected format.
-	ErrTorInvalidProxyResponse = Err("ErrTorInvalidProxyResponse")
+	ErrTorInvalidProxyResponse = ErrorKind("ErrTorInvalidProxyResponse")
 
 	// ErrTorUnrecognizedAuthMethod indicates the authentication method
 	// provided is not recognized.
-	ErrTorUnrecognizedAuthMethod = Err("ErrTorUnrecognizedAuthMethod")
+	ErrTorUnrecognizedAuthMethod = ErrorKind("ErrTorUnrecognizedAuthMethod")
 
 	// ErrTorGeneralError indicates a general tor error.
-	ErrTorGeneralError = Err("ErrTorGeneralError")
+	ErrTorGeneralError = ErrorKind("ErrTorGeneralError")
 
 	// ErrTorNotAllowed indicates tor connections are not allowed.
-	ErrTorNotAllowed = Err("ErrTorNotAllowed")
+	ErrTorNotAllowed = ErrorKind("ErrTorNotAllowed")
 
 	// ErrTorNetUnreachable indicates the tor network is unreachable.
-	ErrTorNetUnreachable = Err("ErrTorNetUnreachable")
+	ErrTorNetUnreachable = ErrorKind("ErrTorNetUnreachable")
 
 	// ErrTorHostUnreachable indicates the tor host is unreachable.
-	ErrTorHostUnreachable = Err("ErrTorHostUnreachable")
+	ErrTorHostUnreachable = ErrorKind("ErrTorHostUnreachable")
 
 	// ErrTorConnectionRefused indicates the tor connection was refused.
-	ErrTorConnectionRefused = Err("ErrTorConnectionRefused")
+	ErrTorConnectionRefused = ErrorKind("ErrTorConnectionRefused")
 
 	// ErrTorTTLExpired indicates the tor request Time-To-Live (TTL) expired.
-	ErrTorTTLExpired = Err("ErrTorTTLExpired")
+	ErrTorTTLExpired = ErrorKind("ErrTorTTLExpired")
 
 	// ErrTorCmdNotSupported indicates the tor command is not supported.
-	ErrTorCmdNotSupported = Err("ErrTorCmdNotSupported")
+	ErrTorCmdNotSupported = ErrorKind("ErrTorCmdNotSupported")
 
 	// ErrTorAddrNotSupported indicates the tor address type is not supported.
-	ErrTorAddrNotSupported = Err("ErrTorAddrNotSupported")
+	ErrTorAddrNotSupported = ErrorKind("ErrTorAddrNotSupported")
 )
+
+// Error satisfies the error interface and prints human-readable errors.
+func (e ErrorKind) Error() string {
+	return string(e)
+}
+
+// Is implements the interface to work with the standard library's errors.Is.
+//
+// It returns true in the following cases:
+// - The target is an Error and any error in its chain matches the error kind
+// - The target is an ErrorKind and the error kinds match
+func (e ErrorKind) Is(target error) bool {
```

Comment:
This seems backwards to me.  errors.Is is not intended to have a commutative property (`errors.Is(err1, err2)` does not always necessarily equal `errors.Is(err2, err1)`).  If my error variable is type ErrorKind, and I am comparing it to a more-specific target error of type Error, I expect that to always return false, because ErrorKind doesn't wrap anything and cannot equal the target. However, reversing the operands should return true (assuming the ErrorKind strings are equal).

Reply:
In general, I agree.  I'm not sure I 100% agree here, though.  The implementation here is not really unwrapping anything in the `ErrorKind`, rather it is treating a given `ErrorKind` as equivalent to an `Error` with that same given `ErrorKind`.

Without this, if the code returns a given `ErrorKind` directly with no context in one scenario and then with context in another scenario, `errors.Is` would fail to detect that it is what you expect.

For example, you can see the test failures if you remove it:

```
TestErrorKindIsAs: error_test.go:140: ErrDialNil == Error.ErrDialNil: incorrect error identification -- got false, want true
TestErrorKindIsAs: error_test.go:140: ErrDialNil == Error.Err{Error.ErrDialNil}: incorrect error identification -- got false, want true
```

Reply:
I saw the test cases there, but I think they are checking for incorrect behavior.

> Without this, if the code returns a given ErrorKind directly with no context in one scenario and then with context in another scenario, errors.Is would fail to detect that it is what you expect.

This is only if your target error is the wrapping Error type, which you'd probably be constructing yourself if you are going to be using it as a target error with errors.Is, and really more of a bug at the errors.Is call site.  It should probably be using ErrorKind as the target type exclusively for detecting any of the errors returned by this package.

Reply:
This is consistent with how the wrapping errors returned by the fmt package behave: https://play.golang.org/p/9VIQjQnPMcp

Reply:
It's also consistent with how errors.Is behaves with wallet's error types.  There wasn't a test case already for this, so I added one:

```
diff a5c2b1218fc458bd9f63416dc0d5a30ae50999c0 /home/jrick/src/dcrwallet
blob - 715f8bb7213e93b2f9a074a3c4c13c6a09a2d5de
file + errors/errors_test.go
--- errors/errors_test.go
+++ errors/errors_test.go
@@ -67,6 +67,9 @@ func TestIs(t *testing.T) {
        if Is(e, E(Invalid)) {
                t.Fatal("match on different kind")
        }
+       if std.Is(base, E(Permission)) { // std is alias for stdlib errors
+               t.Fatal("match of cause with wrapping target")
+       }
 }
 
 func TestCause(t *testing.T) {
```

With this extra check, the tests still pass, because errors.Is does not consider a wrapping target error to match the underlying base error.  it's also worth noting that only the wrapping Error type from that package implements the Is method, not the Kind type.

Reply:
Alright, I'll remove it.  so long as the code only ever returns errors with context, as it really should anyway, and the caller only ever uses the more specific type as intended, there won't be any issue.

Reply:
Done.